src/solvers/flattening/boolbv_member.cpp

```diff
@@ -20,67 +21,49 @@ bvt boolbvt::convert_member(const member_exprt &expr)
 
   const bvt &struct_bv=convert_bv(struct_op);
 
-  if(struct_op_type.id()==ID_union)
-  {
-    return convert_bv(
-      byte_extract_exprt(byte_extract_id(),
-                         struct_op,
-                         from_integer(0, index_type()),
-                         expr.type()));
-  }
-  else if(struct_op_type.id()==ID_struct)
-  {
-    const irep_idt &component_name=expr.get_component_name();
-    const struct_typet::componentst &components=
-      to_struct_type(struct_op_type).components();
+  const irep_idt &component_name = expr.get_component_name();
+  const struct_union_typet::componentst &components =
+    to_struct_union_type(struct_op_type).components();
 
-    std::size_t offset=0;
+  std::size_t offset = 0;
 
-    for(struct_typet::componentst::const_iterator
-        it=components.begin();
-        it!=components.end();
-        it++)
-    {
-      const typet &subtype=it->type();
-      std::size_t sub_width=boolbv_width(subtype);
+  for(struct_union_typet::componentst::const_iterator it = components.begin();
+      it != components.end();
+      ++it)
+  {
+    const typet &subtype = it->type();
+    const std::size_t sub_width = boolbv_width(subtype);
 
-      if(it->get_name()==component_name)
+    if(it->get_name() == component_name)
+    {
+      if(!base_type_eq(subtype, expr.type(), ns))
       {
-        if(!base_type_eq(subtype, expr.type(), ns))
-        {
-          #if 0
-          std::cout << "DEBUG " << expr.pretty() << "\n";
-          #endif
-
-          error().source_location=expr.find_source_location();
-          error() << "member: component type does not match: "
-                  << subtype.pretty() << " vs. "
-                  << expr.type().pretty() << eom;
-          throw 0;
-        }
-
-        bvt bv;
-        bv.resize(sub_width);
-        assert(offset+sub_width<=struct_bv.size());
-
-        for(std::size_t i=0; i<sub_width; i++)
-          bv[i]=struct_bv[offset+i];
-
-        return bv;
+#if 0
+        std::cout << "DEBUG " << expr.pretty() << "\n";
+#endif
+
+        error().source_location = expr.find_source_location();
+        error() << "member: component type does not match: " << subtype.pretty()
+                << " vs. " << expr.type().pretty() << eom;
+        throw 0;
```

Comment:
I think this is an invariant, as it's our responsibility to make sure this is true, not a user error